File: base/doc/guides/SecondaryProducerExample.cpp
#include "rgma/RGMAException.h"
#include "rgma/RGMAService.h"
#include "rgma/SecondaryProducer.h"
#include "rgma/Storage.h"
#include "rgma/SupportedQueries.h"
#include "rgma/TimeInterval.h"
#include "rgma/TimeUnit.h"

#include <string>
#include <iostream>

using glite::rgma::RGMAException;
using glite::rgma::RGMAService;
using glite::rgma::SecondaryProducer;
using glite::rgma::Storage;
using glite::rgma::SupportedQueries;
using glite::rgma::TimeInterval;
using glite::rgma::TimeUnit;

int main(int argc, char* argv[])
{

   SecondaryProducer* sp = 0;

   try
   {
      std::string location("cppExample");
      Storage storage = Storage(location);
      sp = new SecondaryProducer(storage, SupportedQueries::CL);

      std::string predicate("");
      TimeInterval historyRetentionPeriod(2, TimeUnit::HOURS);
      sp->declareTable("default.userTable", predicate, historyRetentionPeriod);

      TimeInterval terminationInterval = RGMAService::getTerminationInterval();

      while(true)
      {
         sp->showSignOfLife();
         sleep(terminationInterval.getValueAs(TimeUnit::SECONDS) / 3);
      }

   }
   catch (RGMAException e)
   {
      std::cerr << "ERROR: "<< e.getMessage() << std::endl;
   }
   catch (...)
   {
      std::cerr << "ERROR: Cuaght unexpected error" << std::endl;
   }

   if (sp)
   {
      try
      {
         sp->close();
      }
      catch (RGMAException e)
      {
      }
   }

   delete sp;
   sp = 0;

}

File: base/doc/guides/SecondaryProducerExample.java
import org.glite.rgma.RGMAException;
import org.glite.rgma.RGMAService;
import org.glite.rgma.SecondaryProducer;
import org.glite.rgma.Storage;
import org.glite.rgma.SupportedQueries;
import org.glite.rgma.TimeInterval;
import org.glite.rgma.TimeUnit;

public class SecondaryProducerExample {
  public static void main(String[] args) {

    SecondaryProducer secondaryProducer = null;

    try {
      String location = "javaExample";
      Storage storage = Storage.getDatabaseStorage(location);
      secondaryProducer = new SecondaryProducer(storage,
          SupportedQueries.CL);

      String predicate = "";
      TimeInterval historyRetentionPeriod = new TimeInterval(2,
          TimeUnit.HOURS);
      secondaryProducer.declareTable("default.userTable", predicate,
          historyRetentionPeriod);

      TimeInterval terminationInterval = RGMAService
          .getTerminationInterval();

      while (true) {
        secondaryProducer.showSignOfLife();
        try {
          Thread.sleep((terminationInterval
              .getValueAs(TimeUnit.SECONDS) / 3) * 1000);
        } catch (InterruptedException e) {
        }
      }

    } catch (RGMAException e) {
      System.err.println("ERROR: " + e.toString());
      System.exit(1);

    } finally {
      if (secondaryProducer != null) {
        try {
          secondaryProducer.close();
        } catch (RGMAException e) {
        }
      }
    }

  }
}
